DisCatSharp/Clients/DiscordClient.EventHandlers.cs

```diff
@@ -0,0 +1,202 @@
+		/// Perfectly mirrors <see cref="RegisterEventHandler(object, bool)"/>.
+		/// </summary>
+		/// <param name="handler"></param>
+		/// <param name="wasRegisteredWithStatic"></param>
+		public void UnregisterEventHandler(object handler, bool wasRegisteredWithStatic = false)
+			=> this.UnregisterEventHandlerImpl(handler, handler.GetType(), wasRegisteredWithStatic);
+
+		/// <summary>
+		/// Perfectly mirrors <see cref="RegisterStaticEventHandler(Type)"/>.
+		/// </summary>
+		/// <param name="t"></param>
+		public void UnregisterStaticEventHandler(Type t) => this.UnregisterEventHandlerImpl(null, t);
+
+		/// <summary>
+		/// Perfectly mirrors <see cref="RegisterStaticEventHandler{T}()"/>.
+		/// </summary>
+		/// <typeparam name="T"></typeparam>
+		public void UnregisterStaticEventHandler<T>() => this.UnregisterEventHandler(typeof(T));
+
+		/// <summary>
+		/// Perfectly mirrors <see cref="RegisterEventHandler(Type)"/>.
+		/// </summary>
+		/// <param name="t"></param>
+		public void UnregisterEventHandler(Type t)
+		{
+			if (t.IsAbstract)
+				this.UnregisterStaticEventHandler(t);
+			else
+			{
+				if (!this._typeToAnonymousHandlers.TryGetValue(t, out var anonObjs)
+					|| anonObjs.Count == 0)
+					return; // Wasn't registered
+
+				object anon = anonObjs[0];
+				anonObjs.RemoveAt(0);
+
+				if (anonObjs.Count == 0)
+					this._typeToAnonymousHandlers.Remove(t);
+
+				this.UnregisterEventHandlerImpl(anon, t);
+			}
+
+
+		}
+
+		/// <summary>
+		/// Perfectly mirrors <see cref="RegisterEventHandler{T}()"/>.
+		/// </summary>
+		/// <typeparam name="T"></typeparam>
+		public void UnregisterEventHandler<T>() => this.UnregisterEventHandler(typeof(T));
+
+		/// <summary>
+		/// Perfectly mirrors <see cref="RegisterEventHandlers(Assembly)"/>.
+		/// </summary>
+		/// <param name="assembly"></param>
+		public void UnregisterEventHandlers(Assembly assembly)
+		{
+			foreach (Type t in GetEventHandlersFromAssembly(assembly))
+				this.UnregisterEventHandler(t);
+		}
+
+		private static IEnumerable<Type> GetEventHandlersFromAssembly(Assembly assembly)
+			=> assembly.GetTypes()
+				.Where(t => t.GetCustomAttribute<EventHandler>() is not null);
+
+		private void UnregisterEventHandlerImpl(object? handler, Type type, bool registerStatic = true)
+		{
+			if (!this._registrationToDelegate.TryGetValue((handler, type, registerStatic), out var delegateLists)
+				|| delegateLists.Count == 0)
+				return;
+
+			foreach (var (evnt, dlgt) in delegateLists[0])
+				evnt.RemoveEventHandler(this, dlgt);
+
+			delegateLists.RemoveAt(0);
+			if (delegateLists.Count == 0)
+				this._registrationToDelegate.Remove((handler, type, registerStatic));
+		}
+
+		private void RegisterEventHandlerImpl(object? handler, Type type, bool registerStatic = true)
+		{
+			var delegates = (
+				from method in type.GetMethods(BindingFlags.Instance | BindingFlags.Static | BindingFlags.Public | BindingFlags.NonPublic)
+				let attribute = method.GetCustomAttribute<Event>()
```

Comment:
Room for discussion here.

Should we also support mapping these methods to enums? To prevent typos as strings, and to easily discover the type of events a user can actually use?

So something like

[Event(DiscordEvents.MessageCreated)]

Reply:
Yes please

Reply:
Added the enum, removed the string constructor and added relevant tests to maintain consistency between the events and the enum.